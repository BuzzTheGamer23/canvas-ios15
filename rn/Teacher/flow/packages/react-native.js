// @flow
export type PanResponderGestureState = {
  stateID: number,
  moveX: number,
  moveY: number,
  x0: number,
  y0: number,
  dx: number,
  dy: number,
  vx: number,
  vy: number,
}
