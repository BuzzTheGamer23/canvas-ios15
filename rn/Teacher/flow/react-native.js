//
// Copyright (C) 2017-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

// @flow
/* eslint-disable */

// define basic style prop types
// the internal ones are the best we have
import {
  type ____ViewStyleProp_Internal,
  type ____TextStyleProp_Internal,
  type ____ImageStyleProp_Internal,
} from 'react-native/Libraries/StyleSheet/StyleSheetTypes'

declare type ViewStyleProp = ____ViewStyleProp_Internal
declare type TextStyleProp = ____TextStyleProp_Internal
declare type ImageStyleProp = ____ImageStyleProp_Internal
