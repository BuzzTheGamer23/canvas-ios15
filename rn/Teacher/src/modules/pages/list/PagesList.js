//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

/* @flow */

import React, { Component } from 'react'
import { connect } from 'react-redux'
import {
  View,
  StyleSheet,
  FlatList,
} from 'react-native'
import i18n from 'format-message'

import Screen from '../../../routing/Screen'
import Row from '../../../common/components/rows/Row'
import Actions from './actions'
import Images from '../../../images'
import RowSeparator from '../../../common/components/rows/RowSeparator'
import { getPages } from '../../../canvas-api'
import { alertError } from '../../../redux/middleware/error-handler'
import PublishedIcon from '../../../common/components/PublishedIcon'
import { Text } from '../../../common/text'
import ListEmptyComponent from '../../../common/components/ListEmptyComponent'

type StateProps = AsyncState & {
  pages: Page[],
  courseName: string,
  courseColor: ?string,
}

type OwnProps = {
  courseID: string,
}

export type Props = OwnProps & StateProps & typeof Actions & NavigationProps

export class PagesList extends Component<Props, any> {
  static defaultProps = {
    getPages,
  }

  constructor (props: Props) {
    super(props)

    this.state = {
      pending: false,
    }
  }

  componentWillMount () {
    this.refresh()
  }

  render () {
    return (
      <Screen
        navBarStyle='dark'
        title={i18n('Pages')}
        subtitle={this.props.courseName}
      >
        <View style={styles.container}>
          <FlatList
            data={this.props.pages}
            renderItem={this.renderRow}
            keyExtractor={(item, index) => item.url}
            testID='pages.list.list'
            refreshing={this.state.pending}
            onRefresh={this.refresh}
            ItemSeparatorComponent={RowSeparator}
            ListEmptyComponent={
              this.state.pending && !this.props.refreshing ? null
              : <ListEmptyComponent title={i18n('There are no pages to display.')} />
            }
          />
        </View>
      </Screen>
    )
  }

  renderRow = ({ item, index }: { item: Page, index: number }) => {
    let icon = (
      <View style={styles.rowIcon}>
        <PublishedIcon published={item.published} tintColor={this.props.courseColor} image={Images.course.pages} />
      </View>
    )
    return (
      <Row
        title={item.title}
        subtitle={i18n("{ date, date, 'MMM d'} at { date, time, short }", { date: new Date(item.created_at) })}
        border='bottom'
        height='auto'
        disclosureIndicator={true}
        testID={`pages.list.page.row-${index}`}
        onPress={this.selectPage(item)}
        renderImage={() => icon}
        accessories={item.front_page && (
          <View style={styles.rowFrontPagePill}>
            <Text style={styles.rowFrontPagePillText} testID='pages.list.front-page.pill'>Front Page</Text>
          </View>
        )}
      />
    )
  }

  selectPage (page: Page) {
    const route = `/courses/${this.props.courseID}/pages/${page.url}`
    return () => this.props.navigator.show(route, { modal: false })
  }

  refresh = async () => {
    this.setState({ pending: true })
    try {
      const { data } = await this.props.getPages(this.props.courseID)
      this.props.refreshedPages(data, this.props.courseID)
    } catch (error) {
      alertError(error)
    }
    this.setState({ pending: false })
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  rowIcon: {
    alignSelf: 'flex-start',
  },
  rowFrontPagePill: {
    paddingHorizontal: 8,
    paddingVertical: 2,
    backgroundColor: '#73818C',
    borderRadius: 4,
  },
  rowFrontPagePillText: {
    color: 'white',
    fontSize: 14,
  },
})

export function mapStateToProps ({ entities }: AppState, { courseID }: OwnProps): StateProps {
  let pages = []
  let courseName = ''
  let courseColor = null
  if (entities &&
    entities.courses &&
    entities.courses[courseID] &&
    entities.courses[courseID].pages) {
    const course = entities.courses[courseID]
    const refs = course.pages.refs
    if (course.course) {
      courseName = course.course.name
      courseColor = course.color
    }
    pages = refs
      .map(ref => entities.pages[ref].data)
      .sort((a, b) => a.title.toLowerCase() > b.title.toLowerCase() ? 1 : -1)
  }
  return {
    pages,
    courseName,
    courseColor,
  }
}

const Connected = connect(mapStateToProps, Actions)(PagesList)
export default (Connected: Component<Props, any>)
