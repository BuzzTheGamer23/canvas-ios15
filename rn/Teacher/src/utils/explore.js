/* @flow */

function hasProp (object: any, key: string, value: any): boolean {
  if (object instanceof Object) {
    const props = object.props
    if (props && props instanceof Object) {
      for (var prop in props) {
        if (prop === key && props[prop] === value) {
          return true
        }
      }
    }
  }

  return false
}

function selectByProp (object: any, propKey: string, value: any): ?any {
  if (!(typeof object === 'object')) {
    return null
  }

  if (hasProp(object, propKey, value)) {
    return object
  }

  let result = null

  for (var key in object) {
    if (object[key] instanceof Array) {
      object[key].forEach((e) => {
        result = selectByProp(e, propKey, value)
        if (result) {
          return result
        }
      })
    }

    result = selectByProp(object[key], propKey, value)
    if (result) {
      return result
    }
  }

  return result
}

class ComponentExplorer {
  component: any

  constructor (component: any) {
    this.component = component
  }

  selectByProp (propKey: string, value: any): ?any {
    return selectByProp(this.component, propKey, value)
  }

  selectByID (id: string): ?any {
    return this.selectByProp('testID', id)
  }
}

export default function explore (component: any): ComponentExplorer {
  return new ComponentExplorer(component)
}
