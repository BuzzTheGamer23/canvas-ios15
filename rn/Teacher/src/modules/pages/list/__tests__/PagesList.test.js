//
// Copyright (C) 2016-present Instructure, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

/* @flow */

import React from 'react'
import 'react-native'
import renderer from 'react-test-renderer'
import { PagesList, mapStateToProps, type Props } from '../PagesList'
import explore from '../../../../../test/helpers/explore'

jest
  .mock('Button', () => 'Button')
  .mock('TouchableHighlight', () => 'TouchableHighlight')
  .mock('TouchableOpacity', () => 'TouchableOpacity')

const template = {
  ...require('../../../../__templates__/page'),
  ...require('../../../../__templates__/helm'),
  ...require('../../../../redux/__templates__/app-state'),
}

describe('PagesList', () => {
  let props: Props
  beforeEach(() => {
    props = {
      pages: [template.page()],
      courseID: '1',
      getPages: jest.fn(() => Promise.resolve({ data: [template.page()] })),
      refreshedPages: jest.fn(),
      navigator: template.navigator(),
    }
  })

  it('renders', () => {
    expect(render(props).toJSON()).toMatchSnapshot()
  })

  it('renders empty', () => {
    props.pages = []
    expect(render(props)).toMatchSnapshot()
  })

  it('refreshes on mount', () => {
    const spy = jest.fn(() => Promise.resolve({ data: [template.page()] }))
    props.getPages = spy
    const view = render(props)
    view.getInstance().componentWillMount()
    expect(spy).toHaveBeenCalledWith(props.courseID)
  })

  it('dispatches refreshedPages action on refresh', async () => {
    const pages = [template.page()]
    props.getPages = jest.fn(() => Promise.resolve({ data: pages }))
    const spy = jest.fn()
    props.refreshedPages = spy
    const view = render(props)
    const list: any = explore(view.toJSON()).selectByType('RCTScrollView')
    await list.props.onRefresh()
    expect(spy).toHaveBeenCalledWith(pages, props.courseID)
  })

  it('renders front page pill', () => {
    props.pages = [template.page({ front_page: true })]
    const pill = explore(render(props).toJSON()).selectByID('pages.list.front-page.pill')
    expect(pill).not.toBeNull()
  })

  it('navigates to page details', () => {
    const spy = jest.fn()
    props.navigator = template.navigator({
      show: spy,
    })
    props.pages = [template.page({ url: 'abc' })]
    props.courseID = '1'
    const view = render(props)
    const row: any = explore(view.toJSON()).selectByID('pages.list.page.row-0')
    row.props.onPress()
    expect(spy).toHaveBeenCalledWith('/courses/1/pages/abc', { modal: false })
  })

  function render (props: Props, options: any = {}): any {
    return renderer.create(<PagesList {...props} />, options)
  }
})

describe('mapStateToProps', () => {
  it('maps state to props', () => {
    const one = template.page({ url: 'page-1', title: 'A' })
    const two = template.page({ url: 'page-2', title: 'B' })
    const three = template.page({ url: 'page-3', title: 'C' })
    const state = template.appState({
      entities: {
        courses: {
          '1': {
            color: '#fff',
            course: {
              name: 'Course 1',
            },
            pages: {
              pending: 0,
              refs: ['page-1', 'page-3'],
            },
          },
        },
        pages: {
          'page-1': { data: one },
          'page-2': { data: two },
          'page-3': { data: three },
        },
      },
    })
    expect(mapStateToProps(state, { courseID: '1' })).toEqual({
      pages: [one, three],
      courseName: 'Course 1',
      courseColor: '#fff',
    })
  })

  it('maps empty state to props', () => {
    const state = template.appState()
    expect(mapStateToProps(state, { courseID: '1' })).toEqual({
      pages: [],
      courseName: '',
      courseColor: null,
    })
  })

  it('sorts pages alphabetically by title', () => {
    const A = template.page({ url: 'A', title: 'A' })
    const b = template.page({ url: 'b', title: 'b' })
    const C = template.page({ url: 'C', title: 'C' })
    const state = template.appState({
      entities: {
        courses: {
          '1': {
            color: '#fff',
            course: {
              name: 'Course 1',
            },
            pages: {
              pending: 0,
              refs: ['A', 'b', 'C'],
            },
          },
        },
        pages: {
          'A': { data: A },
          'b': { data: b },
          'C': { data: C },
        },
      },
    })
    expect(mapStateToProps(state, { courseID: '1' })).toMatchObject({
      pages: [A, b, C],
    })
  })
})
