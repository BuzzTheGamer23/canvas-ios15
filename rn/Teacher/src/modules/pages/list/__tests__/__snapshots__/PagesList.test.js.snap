// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PagesList renders 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    ListEmptyComponent={null}
    data={
      Array [
        Object {
          "body": "<p>Hello, World!</p>",
          "created_at": "2017-03-17T19:15:25Z",
          "editing_roles": "teachers",
          "front_page": false,
          "hide_from_students": false,
          "published": true,
          "title": "Page 1",
          "updated_at": "2017-03-17T19:15:25Z",
          "url": "page-1",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onRefresh={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    onViewableItemsChanged={undefined}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        progressViewOffset={undefined}
        refreshing={true}
      />
    }
    refreshing={true}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    testID="pages.list.list"
    updateCellsBatchingPeriod={50}
    windowSize={21}
  >
    <RCTRefreshControl />
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <TouchableHighlight
          accessibilityLabel={undefined}
          accessibilityTraits={
            Array [
              "button",
            ]
          }
          onPress={[Function]}
          style={
            Array [
              undefined,
              Object {
                "borderBottomColor": "#C7CDD1",
                "borderBottomWidth": 0.5,
              },
            ]
          }
          testID="pages.list.page.row-0"
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "backgroundColor": "white",
                  "flex": 1,
                  "flexDirection": "row",
                  "minHeight": 54,
                  "paddingBottom": 12,
                  "paddingLeft": 16,
                  "paddingRight": 16,
                  "paddingTop": 12,
                },
                Object {
                  "backgroundColor": "white",
                },
              ]
            }
          >
            <View
              style={
                Object {
                  "alignSelf": "flex-start",
                }
              }
            >
              <View
                accessibilityLabel="Published"
                style={
                  Object {
                    "alignItems": "center",
                    "flex": 0,
                    "height": 32,
                    "width": 46,
                  }
                }
              >
                <Image
                  source={1}
                  style={
                    Array [
                      Object {
                        "position": "absolute",
                      },
                      Object {
                        "tintColor": undefined,
                      },
                    ]
                  }
                />
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "white",
                      "borderRadius": 8,
                      "height": 16,
                      "justifyContent": "center",
                      "left": 20,
                      "position": "absolute",
                      "top": 10,
                      "width": 16,
                    }
                  }
                >
                  <Image
                    source={1}
                    style={
                      Object {
                        "height": 12,
                        "tintColor": "#00AC18",
                        "width": 12,
                      }
                    }
                  />
                </View>
              </View>
            </View>
            <View
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "column",
                  "justifyContent": "center",
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                disabled={false}
                ellipsizeMode="tail"
                numberOfLines={0}
                style={
                  Array [
                    Object {
                      "fontFamily": ".SFUIDisplay",
                    },
                    Object {
                      "color": "#2D3B45",
                      "fontSize": 16,
                    },
                    Array [
                      Object {
                        "fontWeight": "600",
                      },
                      undefined,
                    ],
                    Object {
                      "fontFamily": ".SFUIDisplay-semibold",
                    },
                  ]
                }
              >
                Page 1
              </Text>
              <Text
                accessible={true}
                allowFontScaling={true}
                disabled={false}
                ellipsizeMode="tail"
                style={
                  Array [
                    Object {
                      "fontFamily": ".SFUIDisplay",
                    },
                    Object {
                      "color": "#2D3B45",
                      "fontSize": 16,
                    },
                    Object {
                      "color": "#8B969E",
                      "fontSize": 14,
                      "marginTop": 2,
                    },
                    Object {
                      "fontFamily": ".SFUIDisplay",
                    },
                  ]
                }
                testID="pages.list.page.row-0-subtitle-lbl"
              >
                Mar 17 at 1:15 PM
              </Text>
            </View>
            <View
              style={
                Object {
                  "flex": 0,
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "flexDirection": "row",
                    "justifyContent": "flex-end",
                  }
                }
              >
                <View
                  style={
                    Object {
                      "alignSelf": "center",
                      "backgroundColor": "transparent",
                      "borderColor": "#c7c7cc",
                      "borderRightWidth": 2,
                      "borderTopWidth": 2,
                      "height": 10,
                      "marginLeft": 7,
                      "transform": Array [
                        Object {
                          "rotate": "45deg",
                        },
                      ],
                      "width": 10,
                    }
                  }
                />
              </View>
            </View>
          </View>
        </TouchableHighlight>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`PagesList renders empty 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    ListEmptyComponent={null}
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onRefresh={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    onViewableItemsChanged={undefined}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        progressViewOffset={undefined}
        refreshing={true}
      />
    }
    refreshing={true}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    testID="pages.list.list"
    updateCellsBatchingPeriod={50}
    windowSize={21}
  >
    <RCTRefreshControl />
    <View />
  </RCTScrollView>
</View>
`;
